### PR TITLE
CRUD functionality and test cases optional

### PR DESCRIPTION
**Pull request by omerasifkhan**
1. Provided user option to choose which functionalities they want to create from hygen.
2. Secondly, allowed users to choose, if they want to create test case structure while creating resources.

### CODE REVIEW

File: .hygen/generate/relational-resource/__mock__/mock.ejs.t [hunk 1]
```diff
@@ -1,24 +1,33 @@
...
+<% } %>
+<% if (functionalities.includes('create')) { %>
 import { Create<%= name %>Dto } from '@src/<%= h.inflection.transform(name, ['pluralize', 'underscore', 'dasherize']) %>/dto/create-<%= h.inflection.transform(name, ['underscore', 'dasherize']) %>.dto';
+<% } %>
+<% if (functionalities.includes('update')) { %>
 import { Update<%= name %>Dto } from '@src/<%= h.inflection.transform(name, ['pluralize', 'underscore', 'dasherize']) %>/dto/update-<%= h.inflection.transform(name, ['underscore', 'dasherize']) %>.dto';
+<% } %>
 import { <%= name %> } from '@src/<%= h.inflection.transform(name, ['pluralize', 'underscore', 'dasherize']) %>/domain/<%= h.inflection.transform(name, ['underscore', 'dasherize']) %>';
 // __mock__/<%= h.inflection.transform(name, ['underscore', 'dasherize']) %>.mock.ts
+<% if (functionalities.includes('findAll')) { %>
```
Comment (AhmadHassanButt) [inline .hygen/generate/relational-resource/__mock__/mock.ejs.t:15]: combine the conditional imports of findAll 
**line no 15
line no 4**

File: .hygen/generate/relational-resource/service.spec.ejs.t [hunk 2]
```diff
@@ -1,8 +1,10 @@
...
-to: src/<%= h.inflection.transform(name, ['pluralize', 'underscore', 'dasherize']) %>/<%= h.inflection.transform(name, ['pluralize', 'underscore', 'dasherize']) %>.service.spec.ts
+to: "<%= isAddTestCase ? `src/${h.inflection.transform(name, ['pluralize', 'underscore', 'dasherize'])}/${h.inflection.transform(name, ['pluralize', 'underscore', 'dasherize'])}.service.spec.ts` : null %>"
 ---
 import { Test, TestingModule } from '@nestjs/testing';
-import { paginationOptions, mock<%= name %>,mockCreate<%= name %>Dto, mockUpdate<%= name %>Dto } from './__mock__/<%= h.inflection.transform(name, ['underscore', 'dasherize']) %>.mock';
+<% if (functionalities.includes('findAll')) { %>
+import { paginationOptions, mock<%= name %>, mockCreate<%= name %>Dto, mockUpdate<%= name %>Dto } from './__mock__/<%= h.inflection.transform(name, ['underscore', 'dasherize']) %>.mock';
+<% } %>
 import { <%= h.inflection.transform(name, ['pluralize']) %>Service } from './<%= h.inflection.transform(name, ['pluralize', 'underscore', 'dasherize']) %>.service';
 import { <%= name %>Repository } from './infrastructure/persistence/<%= h.inflection.transform(name, ['underscore', 'dasherize']) %>.repository';
```
Comment (AhmadHassanButt) [inline .hygen/generate/relational-resource/service.spec.ejs.t]: I think this import should be like this :
`import { <%= name %>Repository } from './infrastructure/persistence/<%= h.inflection.transform(name, ['underscore', 'dasherize']) %>.abstract.repository';`

Can you please verify on your end?

**Comment by AhmadHassanButt** [comment]
@omerasifkhan - @dev-ahmadbilal :
Should we rename this file 
`.hygen/generate/relational-resource/infrastructure/persistence/repository.ejs.t`
to 
`.hygen/generate/relational-resource/infrastructure/persistence/abstract.repository.ejs.t`

As an abstract key-word will remain consistent while generating the abstract repository of a resource.

File: .hygen/generate/relational-resource/__mock__/mock.ejs.t (hunk 1 above)
Comment (omerasifkhan) [reply .hygen/generate/relational-resource/__mock__/mock.ejs.t:15]: I think it's fine as line 4 is an import statement and 15 is actual data

File: .hygen/generate/relational-resource/service.spec.ejs.t (hunk 2 above)
Comment (omerasifkhan) [reply .hygen/generate/relational-resource/service.spec.ejs.t]: done updated

File: .hygen/generate/relational-resource/__mock__/mock.ejs.t (hunk 1 above)
Comment (AhmadHassanButt) [reply .hygen/generate/relational-resource/__mock__/mock.ejs.t:15]: got it

**Comment by dev-ahmadbilal** [comment]
> @omerasifkhan - @dev-ahmadbilal : Should we rename this file `.hygen/generate/relational-resource/infrastructure/persistence/repository.ejs.t` to `.hygen/generate/relational-resource/infrastructure/persistence/abstract.repository.ejs.t`
> 
> As an abstract key-word will remain consistent while generating the abstract repository of a resource.

Yes, I think we already did this in our project.

File: .hygen/generate/relational-resource/prompt.js [hunk 3]
```diff
@@ -0,0 +1,20 @@
...
+    type: 'multiselect',
+    name: 'functionalities',
+    message: 'Select the functionalities you want to include:',
+    choices: [
+      { name: 'create', value: 'create' },
+      { name: 'findAll', value: 'findAll' },
+      { name: 'findOne', value: 'findOne' },
+      { name: 'update', value: 'update' },
+      { name: 'delete', value: 'delete' },
+    ],
```
Comment (dev-ahmadbilal) [inline .hygen/generate/relational-resource/prompt.js:18]: Can we have short names like instead of create user just add c.

Comment (omerasifkhan) [reply .hygen/generate/relational-resource/prompt.js:18]: It is basically multi-select user don't have to write it. We can write "c"  but I think it's more ambiguous

**Pull request merged**